File: backend/app/schemas/misp_explorer.py
"""MISP Instance Explorer DTO (spec §4.10)."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class MispFeedRow(BaseModel):
    id: str | None = None
    name: str | None = None
    url: str | None = None
    source_format: str | None = Field(default=None, description="MISP / CSV / freetext / unknown")
    enabled: bool | None = None
    last_fetch: datetime | None = None
    event_count: int | None = Field(default=None, description="Best-effort; may be unknown")
    live_sync: bool | None = Field(default=None, description="Enabled feed / scheduling hint")
    cache_age_seconds: int | None = Field(default=None, description="Seconds since last_fetch if known")


class MispServerRow(BaseModel):
    id: str | None = None
    name: str | None = None
    url: str | None = None
    push: bool | None = None
    pull: bool | None = None
    last_sync: datetime | None = None
    sync_status: str | None = Field(default=None, description="success | failed | never | unknown")
    event_count: int | None = None


class MispTaxonomyRow(BaseModel):
    namespace: str | None = None
    enabled: bool | None = None
    description: str | None = None


class MispStatsPanel(BaseModel):
    total_events: int | None = None
    total_attributes: int | None = None
    total_objects: int | None = None
    feeds_configured: int | None = None
    feeds_enabled: int | None = None
    connected_servers: int | None = None
    misp_version: str | None = None
    last_event_added: datetime | None = None


class MispExplorerResponse(BaseModel):
    connected: bool = True
    base_url: str
    resolution: str = Field(
        default="user",
        description="user | platform_fallback | inline_test",
    )
    misp_version: str | None = None
    feeds: list[MispFeedRow] = Field(default_factory=list)
    servers: list[MispServerRow] = Field(default_factory=list)
    taxonomies: list[MispTaxonomyRow] = Field(default_factory=list)
    stats: MispStatsPanel = Field(default_factory=MispStatsPanel)
    sync_indicator: str = Field(default="idle", description="syncing | idle | error")
    source_errors: dict[str, str] = Field(default_factory=dict)
    fetched_at: datetime
    raw_version: dict[str, Any] | None = Field(default=None, exclude=True)

File: frontend/lib/types/misp-explorer.ts
export type MispFeedRow = {
  id: string | null;
  name: string | null;
  url: string | null;
  source_format: string | null;
  enabled: boolean | null;
  last_fetch: string | null;
  event_count: number | null;
  live_sync: boolean | null;
  cache_age_seconds: number | null;
};

export type MispServerRow = {
  id: string | null;
  name: string | null;
  url: string | null;
  push: boolean | null;
  pull: boolean | null;
  last_sync: string | null;
  sync_status: string | null;
  event_count: number | null;
};

export type MispTaxonomyRow = {
  namespace: string | null;
  enabled: boolean | null;
  description: string | null;
};

export type MispStatsPanel = {
  total_events: number | null;
  total_attributes: number | null;
  total_objects: number | null;
  feeds_configured: number | null;
  feeds_enabled: number | null;
  connected_servers: number | null;
  misp_version: string | null;
  last_event_added: string | null;
};

export type MispExplorerResponse = {
  connected: boolean;
  base_url: string;
  resolution: string;
  misp_version: string | null;
  feeds: MispFeedRow[];
  servers: MispServerRow[];
  taxonomies: MispTaxonomyRow[];
  stats: MispStatsPanel;
  sync_indicator: string;
  source_errors: Record<string, string>;
  fetched_at: string;
};
